src/main/java/uk/co/ramp/EvaluateCase.java

```diff
@@ -0,0 +1,230 @@
+                LOGGER.trace("user {} has requested a test. Current Status: {}-{}", p.id(), p.status(), p.alertStatus());
+                break;
+            case NONE:
+                LOGGER.trace("user {} has been alerted. Current Status: {}-{}", p.id(), p.status(), p.alertStatus());
+                updateAlertStatus(ALERTED, time);
+                break;
+        }
+
+        return new HashSet<>();
+    }
+
+    private Set<Integer> alertAllContacts() {
+
+        Set<Integer> contactIds = new HashSet<>();
+        // add all ids
+        for (ContactRecord r : p.contactRecords()) {
+            contactIds.add(r.from());
+            contactIds.add(r.to());
+        }
+
+        // remove self
+        contactIds.remove(p.id());
+
+        LOGGER.trace("Alerting " + contactIds);
+        return contactIds;
+
+    }
+
+
+    private void updateAlertStatus(final AlertStatus newStatus, final int currentTime) {
+
+        int time = getDistributionValue(properties.meanTestTime(), properties.maxTestTime());
+        AlertStatus oldstatus = p.alertStatus();
+        p.setAlertStatus(newStatus);
+        p.setNextAlertStatusChange(currentTime + time);
+        LOGGER.trace("id: {} alertStatus: {} -> {} at t={}. Current Virus status: {}", p.id(), oldstatus, newStatus, currentTime, p.status());
+
+    }
+
+    private VirusStatus determineOutcome(Case p) {
+        //TODO add real logic
+        return p.health() > rng.nextUniform(0, 1) ? RECOVERED : DEAD;
+    }
+
+    private VirusStatus determineInfection(Case p) {
+        //TODO add real logic
+        return p.health() > rng.nextUniform(0, 1) ? INFECTED : INFECTED_SYMP;
+    }
+
+    public void randomExposure(int t) {
+        p.setExposedBy(-2);
+        p.setExposedTime(t);
+        if (p.status() == SUSCEPTIBLE) {
+            LOGGER.trace("Person with id: {} has been randomly exposed at time {}", p.id(), t);
+            updateVirusStatus(EXPOSED, t);
+        } else {
+            String message = String.format("The person with id: %d should not be able to transition from %s to %s", p.id(), p.status(), EXPOSED);
+            LOGGER.error(message);
+            throw new InvalidStatusTransitionException(message);
+        }
+    }
+
+    int getDistributionValue(double mean, double max) {
+
+        int value = (int) Math.round(mean);
+        double sample;
+        switch (properties.progressionDistribution()) {
+            case GAUSSIAN:
+                sample = rng.nextGaussian(mean, mean / 2d);
+                break;
+            case LINEAR:
+                sample = rng.nextUniform(mean - mean / 2d, mean + mean / 2d);
+                break;
+            case EXPONENTIAL:
+                sample = rng.nextExponential(mean);
+                break;
+            case FLAT:
+            default:
+                return value;
+        }
+
+        value = (int) Math.round(sample);
+
+        return Math.min(Math.max(value, 1), (int) max);
```

Comment:
For the exp distribution, setting values which are greater than max to artificially become equal to max causes a large spike in the frequency of value=max being returned. This doesn’t go inline with the exp distribution where the frequency of max should be very low. Some of the other distributions will also have the same issue. We could get around this by recursively calling the method again when this case ( value > max or value < 1) is detected. This would ensure the properties of the distribution are still held whilst not returning back any values > max or < 1. Just a suggestion, maybe there is a better way to resolve this? Happy to hear suggestions or discuss further.

Reply:
That's a very good point. That sounds like a sound resolution too, I'll implement this. 

See before and after:

![Screenshot 2020-05-27 at 10 52 07](https://user-images.githubusercontent.com/64466838/83004985-305ff900-a008-11ea-884b-5384a6574859.png)